[FLINK-24586][table] SQL functions should return STRING instead of VARCHAR(2000)

SteNicholas opened the pull request:
## What is the purpose of the change

There are some SQL functions which currently return VARCHAR(2000). With more strict CAST behavior from FLINK-24413, this could become an issue.

The following functions return VARCHAR(2000) and should be changed to return STRING instead:

* JSON_QUERY
* JSON_OBJECT
* JSON_ARRAY

There are also some more functions which should be evaluated:

* CHR
* REVERSE
* SPLIT_INDEX
* PARSE_URL
* FROM_UNIXTIME
* DECODE

## Brief change log

  - *Updates the return type of the SQL Function `SqlJsonArrayFunction`, `SqlJsonObjectFunction` and `SqlJsonQueryFunction`.*
  - *Updates the return type of the SQL Function `CHR`, `REVERSE`, `SPLIT_INDEX`, `PARSE_URL`, `FROM_UNIXTIME` and `DECODE`.*

## Verifying this change

  - *Updates the `JsonFunctionsITCase` to verfiy the return type of the SQL function `JSON_QUERY`, `JSON_OBJECT` and `JSON_ARRAY` is STRING.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

SteNicholas commented:
@Airblader , could you please take a look at this pull request? Thanks.

slinkydeveloper commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/type/FlinkReturnTypes.java:
```diff
@@ -104,9 +104,9 @@ public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
-    /** Type-inference strategy that always returns "VARCHAR(2000)" with nulls always allowed. */
-    public static final SqlReturnTypeInference VARCHAR_2000_NULLABLE =
-            ReturnTypes.cascade(ReturnTypes.VARCHAR_2000, SqlTypeTransforms.FORCE_NULLABLE);
+    public static final SqlReturnTypeInference VARCHAR_NULLABLE =
```
Perhaps name it `VARCHAR_FORCE_NULLABLE` to make more explicit the nullability behaviour?

SteNicholas replied:
@slinkydeveloper , the `VARCHAR_FORCE_NULLABLE` makes sense to me.

slinkydeveloper commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonQueryFunction.java:
```diff
@@ -0,0 +1,127 @@
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.type.SqlTypeTransforms;
+
+/** This class has been copied from Calcite to replace the return type VARCHAR(2000) with STRING. */
```
I wonder if you can just create a wrapper for `SqlFunction` to avoid copy-pasting `SqlJsonQueryFunction` and `SqlJsonArrayFunction`. If you check the code for `SqlFunction`, you see that the return type inference is used only for:

* `SqlReturnTypeInference getReturnTypeInference()` which is just a getter
* `RelDataType inferReturnType(SqlOperatorBinding opBinding)` which is a wrapper call for SqlReturnTypeInference.inferReturnType`.

WDYT?

SteNicholas replied:
@slinkydeveloper , I agree with the point you mention above. I considered about the wrapper and I'm not sure whether creating a wrapper for `SqlFunction` has no problem. But I will try to create a wrapper for `SqlFunction`.

slinkydeveloper replied:
Thank you! If we can fix it wrapping the class, it's a better solution. Copying classes should always be our last resort :smile:

slinkydeveloper reviewed (changes requested):
Can you also test the return type now is correct for the string functions you changed the return type? `REVERSE`, `CHR`, `SPLIT_INDEX`, `PARSE_URL`, `FROM_UNIXTIME` and `DECODE`?

Also check in `FlinkSqlOperatorTable` for any usages of `ReturnTypes.cascade(ReturnTypes.explicit(SqlTypeName.VARCHAR), SqlTypeTransforms.FORCE_NULLABLE)` and replace them with the new constant

Airblader commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/JsonFunctionsITCase.java:
```diff
@@ -339,7 +339,7 @@ private static TestSpec jsonValueSpec() throws Exception {
                                 null,
                                 STRING(),
-                                VARCHAR(2000)),
+                                STRING()),
```
The last argument can just be omitted if it's the same as the second last one.

slinkydeveloper reviewed (commented):
This is looking good! I left some minor comments, you should also test the return types of REVERSE, CHR, SPLIT_INDEX, PARSE_URL, FROM_UNIXTIME and DECODE. You can add those tests in `StringFunctionsITCase`.

Airblader commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonObjectFunction.java:
```diff
@@ -42,19 +40,18 @@
 /**
- * This class has been copied from Calcite to backport the fix made during CALCITE-4394.
- *
- * <p>TODO Remove this class with Calcite 1.27 and replace it with {@link
```
If not done yet, please also add the TODOs for the Calcite upgrade into the corresponding JIRA issue directly; comments in the code will most likely not be seen.

SteNicholas commented:
@slinkydeveloper , IMO, the previous tests include the return types of REVERSE, CHR, SPLIT_INDEX, PARSE_URL, FROM_UNIXTIME and DECODE because in this pull request I have updated the `testProcessingTimeInnerJoinWithOnClause.out` file. Does it need to add new tests for these functions?

SteNicholas commented:
@slinkydeveloper @Airblader , thanks for the detailed review. I have addressed the above comments. But the tests for `CHR`, `REVERSE`, `SPLIT_INDEX`, `PARSE_URL`, `FROM_UNIXTIME` and `DECODE` couldn't add to `StringFunctionsITCase` which is tested for the `BuiltInFunctionDefinition` not `SqlFunction`.

slinkydeveloper reviewed (approved):
For me we're good to go, once the build is green we can merge it

slinkydeveloper commented:
@flinkbot run azure

SteNicholas commented:
@slinkydeveloper , the failed test has nothing to do with this change. Please check again.

twalthr commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlJsonObjectFunction.java:
```diff
@@ -42,19 +40,21 @@
                 "JSON_OBJECT",
                 SqlKind.OTHER_FUNCTION,
-                ReturnTypes.VARCHAR_2000,
+                VARCHAR_NOT_NULL,
```
introduce a wrapper for this class already. we should not start patching classes copied from Calcite.

twalthr reviewed (changes requested):
Thanks @SteNicholas. One last comment from my side. Otherwise good to merge in the next iteration.

SteNicholas commented:
@twalthr , thanks for the detailed review. I have addressed the last comment above mentioned. Please help to review again.

twalthr reviewed (approved):
LGTM thanks for the update @SteNicholas.

tsreaper commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:
```diff
@@ -1145,13 +1139,13 @@ public boolean isDeterministic() {
     // JSON FUNCTIONS
     public static final SqlFunction JSON_EXISTS = SqlStdOperatorTable.JSON_EXISTS;
     public static final SqlFunction JSON_VALUE = SqlStdOperatorTable.JSON_VALUE;
```
This function is not changed but it is in the list of FLINK-24586. Shall we also change this?

tsreaper reviewed (commented):
Hi all! Sorry for revisiting this PR but I think we might miss `JSON_VALUE`.
